consensus/core/src/synchronizer.rs

```diff
@@ -0,0 +1,480 @@
+                context,
+                commands_receiver,
+                fetch_block_senders,
+            };
+            s.run().await;
+        });
+
+        Arc::new(SynchronizerHandle {
+            commands_sender,
+            tasks: Mutex::new(tasks),
+        })
+    }
+
+    async fn fetch_blocks_from_authority<
+        C: NetworkClient,
+        V: BlockVerifier,
+        D: CoreThreadDispatcher,
+    >(
+        peer_index: AuthorityIndex,
+        network_client: Arc<C>,
+        block_verifier: Arc<V>,
+        context: Arc<Context>,
+        core_dispatcher: Arc<D>,
+        mut receiver: Receiver<BTreeSet<BlockRef>>,
+    ) {
+        const REQUEST_TIMEOUT: Duration = Duration::from_millis(500);
+        const MAX_RETRIES: u32 = 5;
+
+        let mut requests = FuturesUnordered::new();
+
+        loop {
+            tokio::select! {
+                Some(block_refs) = receiver.recv(), if requests.len() < FETCH_BLOCKS_CONCURRENCY => {
+                    requests.push(Self::fetch_blocks_request(network_client.clone(), peer_index, block_refs, REQUEST_TIMEOUT, 1))
+                },
+                Some((response, block_refs, retries)) = requests.next() => {
+                    match response {
+                        Ok(Ok(blocks)) => {
+                            if let Err(err) = Self::process_fetched_blocks(blocks,
+                                peer_index,
+                                block_refs,
+                                core_dispatcher.clone(),
+                                block_verifier.clone(),
+                                context.clone()).await {
+                                error!("Error while processing fetched blocks from peer {peer_index}: {err}");
+                            }
+                        },
+                        Ok(Err(_)) | Err(Elapsed {..}) => {
+                            if retries <= MAX_RETRIES {
+                                requests.push(Self::fetch_blocks_request(network_client.clone(), peer_index, block_refs, REQUEST_TIMEOUT, retries))
+                            } else {
+                                warn!("Max retries {retries} reached while trying to fetch blocks from peer {peer_index}.");
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Processes the requested raw fetched blocks from peer `peer_index`. If no error is returned then
+    /// the verified blocks are immediately sent to Core for processing.
+    async fn process_fetched_blocks<V: BlockVerifier, D: CoreThreadDispatcher>(
+        serialized_blocks: Vec<Bytes>,
+        peer_index: AuthorityIndex,
+        requested_block_refs: BTreeSet<BlockRef>,
+        core_dispatcher: Arc<D>,
+        block_verifier: Arc<V>,
+        context: Arc<Context>,
+    ) -> ConsensusResult<()> {
+        let mut verified_blocks = Vec::new();
+        for serialized_block in serialized_blocks {
```

Comment:
Maybe first make sure serialized_blocks.len() <= requested_block_refs.len()

Reply:
Ah I've got this check on my follow up PR, so it should be there